agent/pkg/podman/podman.go

```diff
@@ -0,0 +1,671 @@
+package podman
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/url"
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	dcontainer "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	"github.com/determined-ai/determined/agent/internal/container"
+	"github.com/determined-ai/determined/agent/internal/options"
+	"github.com/determined-ai/determined/agent/pkg/docker"
+	"github.com/determined-ai/determined/agent/pkg/events"
+	"github.com/determined-ai/determined/master/pkg/aproto"
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/master/pkg/cproto"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/syncx/waitgroupx"
+	"github.com/determined-ai/determined/master/pkg/tasks"
+)
+
+var podmanWrapperEntrypoint = path.Join(tasks.RunDir, tasks.SingularityEntrypointWrapperScript)
+
+const (
+	// TODO(DET-9111): Parameterize this by agent ID.
+	agentTmp         = "/tmp/determined/agent"
+	hostNetworking   = "host"
+	bridgeNetworking = "bridge"
+	envFileName      = "envfile"
+	archivesName     = "archives"
+)
+
+// PodmanContainer captures the state of a container.
+type PodmanContainer struct {
+	PID     int            `json:"pid"`
+	Req     cproto.RunSpec `json:"req"`
+	TmpDir  string         `json:"tmp_dir"`
+	Proc    *os.Process    `json:"-"`
+	Started atomic.Bool    `json:"started"`
+}
+
+// PodmanClient implements ContainerRuntime.
+type PodmanClient struct {
```

Comment:
> I presume most of this is a clone from singularity.go -- couldn't you inherit much of it instead of copy?
> Then just override the differences?

Inheritance in Go seems a bit obtuse. I thought I'd get some feedback from Bradly before slicing into singularity.go

Reply:
I have been researching inheritance in Go. Does seem like as plausible pattern:

https://go.dev/play/p/PnZq3rB2sRd
??

Reply:
Yeah, something like that.   I wasn't thinking it needed to be quite that complex.   For example:

determined-ee/master/internal/rm/resource_manager_iface.go

Then we have common methods in /Users/jerryharrow/git/determined-ee/master/internal/rm/actorrm/actor_resource_manager.go including some "default methods"

Then we override some of them in the actual implementations /Users/jerryharrow/git/determined-ee/master/internal/rm/dispatcherrm/dispatcher_resource_manager.go

Some of this is actor-based, but I don't think the interface and implementations are.

Reply:
That makes sense, as dispatcher RM is defined as:

// DispatcherResourceManager is a resource manager for managing slurm resources.
type DispatcherResourceManager struct {
	*actorrm.ResourceManager
}